### DISCUSSION TITLE
proxy on headless machine

### DISCUSSION THREAD

**Discussion by kdv666** [mitmproxy/mitmproxy#7393] [Q&A]
I want to run mitmproxy on a vps which is a headless machine in the cloud.  I can get it running, and use the curl commands in the tutorial successfully.  My problem is with certificates.  Since that machine is headless, I can't run a browser on it.

To illustrate, my vps in the cloud is drumkit.  My client desktop is sophia.  Both run Debian Linux.

With mitmproxy installed on drumkit, I bring up a console display using ssh, then run the proxy.  I get the properly displayed console screen, and it behaves as the tutorial says, when I use curl commands proxied  to 127.0.0.1:8080.  I tried installing the ca-certificate manually, and no errors were reported.

However if I browse from sophia, telling firefox to use drumkit:8080 as proxy, I get the security error every time. 

I'd appreciate a hint here!   Thanks.

**Comment by Prinzhorn** [comment]
I'm confused. You're saying:

> Since that machine is headless, I can't run a browser on it.

But also

> However if I browse from sophia, telling firefox to use drumkit:8080 as proxy, I get the security error every time.

Are you using Firefox headless? If not, then visit `http://mitm.it`, download the certificate and install it so Firefox can find it. You don't need to do anything on the machine running mitmproxy.

To summarize: your issue has nothing to do with headless. Running mitmproxy (or `mitmdump`) on a server is perfectly normal. You just haven't installed the root certificate properly for Firefox to use.

**Reply by kdv666** [reply]
Many thanks for taking the trouble to reply, I really appreciate it!  And yes, sorry, I can see that I could have been clearer.

Sophia my normal linux desktop, running Debian Mate, and firefox is running on that.

Drunkit is a VPS in the cloud.  It is running Debian with no GUI, i.e. headless.

I installed mitmproxy on Drumkit, listening on 8080 as normal.  Using curl from a shell on drumkit works fine, as is illustrated in the tutorial.

I edited firefox's settings on sophia, entering drumkit:8080 as manual proxy.

The tutorial seems to indicate that the certificate has to be installed on the server running drumkit, because its procedure to install it refers to 127.0.0.1.  That was how I read it, but I am quite possibly wrong.  Should I actually be downloading the certificate onto sophia and installing it there?  I seem to get that impression from what you wrote, so I'll try it and see what happens.  The end goal is that I want to access a particular website using firefox on sophia, but coming from the drumkit IP from the perspective of the website.

Best regards.

Ok, tried doing that on Sophia, then running firefox with the proxy set, but still getting the secure connection failed error.

**Reply by Prinzhorn** [reply]
What tutorial are you referring to? Did you see this on mitm.it?

![image](https://github.com/user-attachments/assets/a0665169-028f-4833-9b76-895f52d07ed9)


> Should I actually be downloading the certificate onto sophia and installing it there?

Yes, that's the client that needs to trust the certificates generated by mitmproxy.

**Reply by kdv666** [reply]
Hello, and thanks once again for taking the trouble to help my rapidly deteriorating brain! 

I was  previously following the Debian tutorial at GitHub.

After reading your answer, I went to mitm.it.  It redirects to docs.mitmproxy.org/stable.

There I went to the "certificates" section, and looked at the "Firefox section.

My problems:

Paragraph 2: "Clicking each of these links will trigger the "trust a new CA" process".

They don't.  They just go to the web page.  Howerver, I did go to the CA manager from settings and added the Mozilla .crt manually.

However, it's still not really working.   Firefox gives me the "unsafe" warning page.  If I click through advanced, on some pages I can click through "advanced", and then "accept the risk", on some I can't. When I do accept the risk, sometimes the pages work, sometimes they don't.  www.google.com doesn't work for example, it only lets me go back.

I may well be missing something, but I can't see it.   Sorry to be a pain!

**Reply by Prinzhorn** [reply]
> After reading your answer, I went to mitm.it. It redirects to docs.mitmproxy.org/stable.

It shouldn't redirect. If you visit it _though the proxy_ you see the instructions and can download the cert.

> I may well be missing something, but I can't see it. Sorry to be a pain!

Our docs might also not have the correct info for your specific setup. I currently don't have any resources to look into this.
